### ISSUE TITLE
Incorrect overridden static method resolution in derived cls

### ISSUE THREAD

**Issue by sourceforge-issue-exporter** [beanshell/beanshell#297]
If you have a derived class that overrides a base
class' static method with an identical signature, and
call the derived method with an object that belongs to
a subtype of the argument type, then bsh calls the base
class'
static method.  \(Does not happen if the actual argument
passed is an exact match for the parameter type.\)

This worked fine with 1.3.0.

Here's the most basic test case:

==== test/Base.java ====
```java
package test;

public class Base {
    public static void mymethod(MyIF s) {
        System.out.println("Base " + s);
    }
}
```
==== test/Derived.java ====
```java
package test;

public class Derived extends Base {
    public static void mymethod(MyIF s) {
        System.out.println("Derived " + s);
    }
}
```
==== test/MyIF.java ====
```java
package test;

interface MyIF {}
```
==== test/MyIFImpl.java ====
```java
package test;

public class MyIFImpl implements MyIF {
    String _s;
    public MyIFImpl(String s) {
        _s = s;
    }
    public String toString() {
        return _s;
    }
}
```
==== test.bsh ====
```java
foo = new test.MyIFImpl("foo");
test.Derived.mymethod(foo); // prints "Base foo"!
```

===================

Reported by: shankarunni

Original Ticket: "beanshell/bugs/159":https://sourceforge.net/p/beanshell/bugs/159

**Comment by nickl-** [comment]
```diff
Index: src/bsh/Reflect.java
===================================================================
--- src/bsh/Reflect.java	(revision 41)
+++ src/bsh/Reflect.java	(working copy)
@@ -726,8 +726,9 @@
 				if ( Types.isSignatureAssignable(
 						idealMatch, targetMatch, round )
 					&& ( (bestMatch == null) ||
-						Types.isSignatureAssignable( targetMatch, bestMatch,
-							Types.JAVA_BASE_ASSIGNABLE )
+						( Types.isSignatureAssignable( targetMatch, bestMatch,
+							Types.JAVA_BASE_ASSIGNABLE ) &&
+						  !Types.areSignaturesEqual(targetMatch, bestMatch) )
 						)
 				)
 				{
Index: src/bsh/Types.java
===================================================================
--- src/bsh/Types.java	(revision 41)
+++ src/bsh/Types.java	(working copy)
@@ -128,6 +128,22 @@
 				throw new InterpreterError("bad case");
 		}
 	}
+	
+	/**
+	 * Are the two signatures exactly equal? This is checked for a special
+	 * case in overload resolution.
+	 */
+	static boolean areSignaturesEqual(Class[] from, Class[] to)
+	{
+		if (from.length != to.length)
+			return false;
+		
+		for (int i = 0; i < from.length; i++)
+			if (from[i] != to[i])
+				return false;
+		
+		return true;
+	}
 
 	private static boolean isSignatureVarargsAssignable(
 		Class[] from, Class[] to )
```

**Comment by nickl-** [comment]
Appears to be resolved in BeanShell 3.0.0-SNAPSHOT.1170

```java
package test;
--> void
interface MyIF {}
--> $0 = interface test.MyIF :Class
public class Base {
    public static void mymethod(MyIF s) {
        System.out.println("Base " + s);
    }
}
--> $1 = class test.Base :Class
public class Derived extends Base {
    public static void mymethod(MyIF s) {
        System.out.println("Derived " + s);
    }
}
--> $2 = class test.Derived :Class

public class MyIFImpl implements MyIF {
    String _s;
    public MyIFImpl(String s) {
        _s = s;
    }
    public String toString() {
        return _s;
    }
}
--> $3 = class test.MyIFImpl :Class
foo = new MyIFImpl("foo");
--> $4 = foo :test.MyIFImpl
Derived.mymethod(foo);
Derived foo
--> void
bsh %
```

Closed: resolved
